### PR TITLE
[Core] Add new token class and protocols

### PR DESCRIPTION
**Pull request by pvaneck**
Here, we add a few new classes to core to support adding/using more data in access tokens received from credentials.

- `AccessTokenInfo`: A new alternative class to `AccessToken` which isn't tuple-based and supports `refresh_on` as well as future extensibility. This naming indicates it's more of a container class for a token and token metadata.
- `SupportsTokenInfo`: A new `Protocol` class. This specifies the method `get_token_info`.
- `TokenRequestOptions` will be a TypeDict that will maintain the optional options for token requests, and will be used in the `get_token_info` method signature. This helps us move away from having to keep adding new kwargs to `get_token`'s method signature every time we need to add new parameters.
- `AsyncSupportsTokenInfo`: The async counterpart to `SupportsTokenInfo`.

Our `BearerTokenCredential` policies will check if a credential has the `get_token_info` method. If so, call it. Otherwise, just call `get_token`.

### CODE REVIEW

File: sdk/core/azure-core/azure/core/credentials.py
Comment (pvaneck) [inline sdk/core/azure-core/azure/core/credentials.py:78]: Opinions on wrapping all optional parameters into a TypedDict like so?

```python
class TokenRequestOptions(TypedDict, total=False):
    """Optional options to use when requesting an expanded token."""

    claims: str
    """Additional claims required in the token, such as those returned in a resource provider's claims
    challenge following an authorization failure."""
    tenant_id: str
    """The tenant ID to include in the token request."""
    enable_cae: bool
    """Indicates whether to enable Continuous Access Evaluation (CAE) for the requested token."""

```

Then can just have the signature be something like:
```python
def get_expanded_token(self, *scopes, options: Optional[TokenRequestOptions] = None, **kwargs)
```

File: sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py
Comment (pvaneck) [inline sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py:101]: Will opt to do a `hasattr` check for "get_expanded_token" since isinstance seems inconsistent in Python 3.8 (and this approach is probably more reliable overall)

File: sdk/core/azure-core/azure/core/credentials.py
Comment (johanste) [reply sdk/core/azure-core/azure/core/credentials.py:78]: What is an implementor of the protocol expected to do if one or more additional_options are unknown?

Comment (pvaneck) [reply sdk/core/azure-core/azure/core/credentials.py]: With a TypedDict maintaining the options, users will get typing errors when trying to pass in unknown options. However, since it's still functionally possible, I've update the docstrings of the protocol methods to explicitly stated that unknown options will be ignored.

Similarly, ignoring certain options is the current pattern with `get_token` since not all credentials support all options.

File: sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py
Comment (xiangyan99) [inline sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py:100]: Should we ways fall back to use get_token? e.g. if user specifies is_pop, we should only try get_token_info.

Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py:100]: Or the plan is we will add that later and identity with pop will depend on 1.32.0?

File: sdk/core/azure-core/azure/core/credentials.py
Comment (xiangyan99) [inline sdk/core/azure-core/azure/core/credentials.py:19]: Let's also add is_pop into the token info.

File: sdk/core/azure-core/azure/core/credentials.py
Comment (xiangyan99) [inline sdk/core/azure-core/azure/core/credentials.py:46]: We don't need to put well-known options into the option bag.

Did I mis-understand?

File: sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py
Comment (pvaneck) [reply sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py:100]: From what I recall, if a token is PoP, then it'll need to be handled differently, likely in a different policy (e.g. `PopTokenCredentialPolicy`). I think the fallback behavior here is fine. Users wanting PoP tokens would need to be using the latest version of identity which would have all credentials implementing `get_token_info`.

File: sdk/core/azure-core/azure/core/credentials.py
Comment (pvaneck) [reply sdk/core/azure-core/azure/core/credentials.py:19]: Should be `token_type` with the default being "bearer", right? (Based on https://github.com/Azure/azure-sdk-for-net/pull/45134/files#diff-fa2c390af9cd1cb54117b35c3ff3d08713ed339adbc4cf0659690f9663558714R46)

I think we can hold off on adding that until we add the rest of the PoP related parameters.

File: sdk/core/azure-core/azure/core/credentials.py
Comment (pvaneck) [reply sdk/core/azure-core/azure/core/credentials.py:46]: I think it makes sense to aggregate all optional token request options in a single container (similar to other languages: [example 1](https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/core/core-auth/src/tokenCredential.ts#L27), [example 2](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azcore/internal/exported/exported.go#L56)), keeping things more organized.

File: sdk/core/azure-core/azure/core/credentials.py
Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/credentials.py:19]: Then we need token_type here, right?

File: sdk/core/azure-core/azure/core/credentials.py
Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/credentials.py:46]: Other languages prefer option bags, but Python prefers flattened arguments, right? By flattening them, we can provide better intellisense.

Comment (pvaneck) [reply sdk/core/azure-core/azure/core/credentials.py:46]: We've established we need an options bag in this case to prevent frequently changing protocol method signatures. And also to prevent issues with implementors of the protocol passing in arbitrary kwargs to transports. The PoP changes appears like it will introduce 3-4 new params. 

If we are putting some optional params in this container class, we might as well put all of them in it. To me, the intellisense experience still works fine since you can still dig down to see what the parameters of the TypedDict are. The majority of SDK users wouldn't be calling this method directly anyway, since it's more of a policy thing.

Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/credentials.py:46]: e.g. if I (user) want to add "transport" in the options, how can I achieve it?

Comment (pvaneck) [reply sdk/core/azure-core/azure/core/credentials.py:46]: We don't support that since that is a Credential constructor level option. New options can be added as-needed if there is a strong enough requirement for them.

Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/credentials.py:46]: Then how about "timeout"? It is supported today.

Comment (pvaneck) [reply sdk/core/azure-core/azure/core/credentials.py:46]: Sure, this will still be supported in `get_token`. 

However, our policies don't propagate these transport level args anyway since we just do a `self._credential.get_token(*self._scopes)` in our `on_request` methods. Our policies are the only users of `get_token_info` , and if for some reason, a user really needs these transport configurations at the token request level, we can always add it later (and maybe in a cleaner way rather than just passing all kwargs). At this stage, I don't think we need it.

Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/credentials.py:46]: Let's take one step back.

What's the user experience of pop token?

Comment (pvaneck) [reply sdk/core/azure-core/azure/core/credentials.py:46]: Outlined in the design gist: https://gist.github.com/christothes/98e0a1c554f180c5ed0c2094d3b93c08#api-changes-overview . Services that support PoP will add the new PopPolicy to their SDK pipelines. I don't think end users of our SDKs will really even notice or have to configure anything for PoP auth. 

The new classes established in this PR can (and will) be extended to support all the new configuration in any case. From our side, it'll be adding the handful of new options to `TokenRequestOptions`, `token_type` to `AccessTokenInfo`, and eventually adding a new pop policy. But again, I want to keep Pop changes separate from this PR.

Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/credentials.py:46]: I was asking for end user experience.

Do you mean user does not need to know if it is pop token or not?

We use pop token as long as it is available?

Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/credentials.py:46]: IOW, the key question is if pop is on by default (as long as it is available).

Comment (pvaneck) [reply sdk/core/azure-core/azure/core/credentials.py:46]: Right, from what I understand from the .NET implementation and the design doc, the PopAuthPolicy will only be added to services that support Pop auth. I believe, right now, only MS Graph supports this? So services that don't work with Pop likely wont even have the Pop policy in their pipelines.

And I believe there should still be a fallback to the BearerTokenCredentialPolicy if for whatever reason that fails. Will need to double check with Chris on that. But that's what I recall.

File: sdk/core/azure-core/azure/core/credentials.py
Comment (xiangyan99) [inline sdk/core/azure-core/azure/core/credentials.py:90]: If I understand correctly, we only need

def get_token_info(self, *scopes: str, options: Optional[dict] = None) -> AccessTokenInfo:

And what is the behavior if we fail to get a token from get_token_info?

Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/credentials.py:90]: And to make sure we are on the same page, right after we ship the new protocol, we need another identity GA with the credentials that implement the new protocol.

Comment (pvaneck) [reply sdk/core/azure-core/azure/core/credentials.py:90]: > def get_token_info(self, *scopes: str, options: Optional[dict] = None) -> AccessTokenInfo:

Putting the TypedDict class in the signature will clue people in to what keys and types are expected in the options dict.
Users can still pass in normal dicts with only a subset of the keys and typing will still be happy if the key names and types align.

```cred.get_token_info("scope", options={"tenant_id": 12345})```

> And what is the behavior if we fail to get a token from get_token_info?

If `get_token_info` is defined and fails, it's treated the same as if `get_token` fails. Also, if `get_token_info` were to fail, likely `get_token` would fail as well.

> And to make sure we are on the same page, right after we ship the new protocol, we need another identity GA with the credentials that implement the new protocol.

Yes, I've been working on the identity side of things separately. We can get the identity protocol implementation changes in as soon as possible. However, things will still continue to work even if we GA this and our credentials don't immediately implement the new `get_token_info` method. The `get_token` method will still continue to be used  in the policies.

File: sdk/core/azure-core/azure/core/credentials.py
Comment (xiangyan99) [inline sdk/core/azure-core/azure/core/credentials.py:40]: Let's also add token_type which is needed for pop.

File: sdk/core/azure-core/azure/core/credentials.py
Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/credentials.py:90]: Then how do we prompt the default value of those arguments?

Comment (pvaneck) [reply sdk/core/azure-core/azure/core/credentials.py:90]: `get_token_info` implementations will treat the absence of an `options` parameter to be None (or its falsey value). Pretty much just treated as if it we were extracting the value from a `kwargs` dict.

With the TypeDict class, all fields can be omitted, but if present, it must be of the specified type.

File: sdk/core/azure-core/azure/core/credentials.py
Comment (pvaneck) [reply sdk/core/azure-core/azure/core/credentials.py:40]: Does it really needed to be added in this PR? PoP is a whole new set of APIs/features where this can be added in that PR.

File: sdk/core/azure-core/azure/core/credentials.py
Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/credentials.py:90]: I meant we could have get_token(*, enable_cae=False, tenant_id=None, use_default_broker_account=False) in the signature so user could clearly see the default value.

If everything in in options and we only have options=None, how can they know the default value of enable_cae in intellisense?

File: sdk/core/azure-core/azure/core/credentials.py
Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/credentials.py:40]: Discussed with Scott, we had not finalized the design yet so we may hold off on the change in policies.

the token_type is relatively a safe change to unblock the pop token.

We still need to confirm the property name and type when Chris is back.

Comment (pvaneck) [reply sdk/core/azure-core/azure/core/credentials.py:45]: Added in latest changeset.

File: sdk/core/azure-core/azure/core/credentials.py
Comment (xiangyan99) [inline sdk/core/azure-core/azure/core/credentials.py:26]: missing keyword token_type?

Comment (pvaneck) [reply sdk/core/azure-core/azure/core/credentials.py:27]: Good catch! Updated

File: sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py
Comment (xiangyan99) [inline sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py:120]: How about having a helper function to handle 

if
call get_token_info
else
call get_token

in a central place?

Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py:120]: The point is get_token does not support pop.

Later we may need to add the logic

if pop
only try get_token_info

If we can put it in a central place, it can increase the maintainability.

File: sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py
Comment (xiangyan99) [inline sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py:43]: Why it is a union?

It can be just AccessTokenInfo, no?

File: sdk/core/azure-core/azure/core/pipeline/policies/_authentication_async.py
Comment (xiangyan99) [inline sdk/core/azure-core/azure/core/pipeline/policies/_authentication_async.py:41]: Same here.

File: sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py
Comment (pvaneck) [reply sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py:43]: The cached token can be either an AccessToken or AccessTokenInfo depending on if `get_token` or `get_token_info` was called.

File: sdk/core/azure-core/azure/core/pipeline/policies/_authentication_async.py
Comment (pvaneck) [reply sdk/core/azure-core/azure/core/pipeline/policies/_authentication_async.py:41]: The cached token can be either an AccessToken or AccessTokenInfo depending on if `get_token` or `get_token_info` was called.

File: sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py
Comment (pvaneck) [reply sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py]: Sure, I centralized the logic to a private method in the respective policy classes.

File: sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py
Comment (kristapratico) [inline sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py:92]: What happens if we let the service validate what's in kwargs?

File: sdk/core/azure-core/azure/core/credentials_async.py
Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/credentials_async.py]: Because credentials_async is public.

We don't want user to use 
from azure.core.credentials_async import AccessToken

Comment (pvaneck) [reply sdk/core/azure-core/azure/core/credentials_async.py]: Pyright/Pylance considers imported symbols [private by default](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md#library-in). So trying to import `AccessToken` would cause it to complain: `"AccessToken" is not exported from module "azure.core.credentials_async"
  Import from "azure.core.credentials" instead`. Would you still prefer to keep the underscore here regardless?

Comment (xiangyan99) [reply sdk/core/azure-core/azure/core/credentials_async.py]: Yes. We cannot assume users all have pyright/pylance enabled. :)

**Review by xiangyan99 — APPROVED** [review]

**Comment by pvaneck** [comment]
CI failures are unrelated to the changes in this PR and have to do with engsys issues that cropped up recently. The failing pipelines _just_ passed yesterday.

**Review by mccoyp — APPROVED** [review]
KV updates LGTM!

**Pull request merged**
